solidity/contracts/hooks/OPL2ToL1CcipReadHook.sol

```diff
@@ -0,0 +1,59 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+pragma solidity >=0.8.0;
+
+// ============ Internal Imports ============
+import {TokenMessage} from "../token/libs/TokenMessage.sol";
+import {StandardHookMetadata} from "./libs/StandardHookMetadata.sol";
+import {AbstractPostDispatchHook} from "./libs/AbstractPostDispatchHook.sol";
+import {AbstractMessageIdAuthHook} from "./libs/AbstractMessageIdAuthHook.sol";
+import {InterchainGasPaymaster} from "../hooks/igp/InterchainGasPaymaster.sol";
+
+/**
+ * @title OPL2ToL1CcipReadHook
+ * @notice Inform an OPL2ToL1ProveWithdrawalIsm that a withdrawal has been executed
+ * on L2
+ * @dev We expect a CCIP-read ISM executing portal.proveWithdrawal() on destination
+ * after 7 days
+ */
+contract OPL2ToL1CcipReadHook is AbstractMessageIdAuthHook {
+    using StandardHookMetadata for bytes;
+    using TokenMessage for bytes;
+
+    // ============ Constants  ============
+    uint32 public constant MIN_GAS_LIMIT = 500_000;
+
+    // ============ Constructor ============
+    constructor(
+        address _mailbox,
+        uint32 _destinationDomain,
+        bytes32 _ism
+    ) AbstractMessageIdAuthHook(_mailbox, _destinationDomain, _ism) {}
+
+    /// @inheritdoc AbstractPostDispatchHook
+    function _quoteDispatch(
+        bytes calldata metadata,
+        bytes calldata message
+    ) internal view override returns (uint256) {
+        return mailbox.quoteDispatch(destinationDomain, ism, message, metadata);
+    }
+
+    /// @inheritdoc AbstractMessageIdAuthHook
+    function _sendMessageId(
+        bytes calldata metadata,
+        bytes calldata message
```

Comment:
It might be useful to name or at least comment the usages of metadata/message here because IIUC, this is the "parent" full Hyperlane message, that we send as the body of the "prove withdrawal" message yeah? Not sure if you talked with @yorhodes about this, but I wonder if its better if we deliniate the message via the body vs. the recipient?

Reply:
You mean wrapping it through `TokenMessage` ? If so, I got it should be done at a higher level (when sending the first message)

Reply:
I actually think its preferable for the hook to be agnostic to the message body

Reply:
Yeah the ISM as the recipient is totally fine, just saying that the differentiation (i.e. `_areWeMessageRecipient`) seems better suited via something in the message body of the extra message for proving. Think of this like a application-specific leading type byte that some folks are doing.

The `message` (i.e. message body) passed in the "extra" message for proving, is right now the underlying full hyperlane message (i.e. the messageBody + the recipient/destination domainParameters). I agree that the hook should be agnostic to the message body, but not sure if it should use that as the body for its (i.e. extra) message itself. You could imagine that it should simply by something like `"proving message" + message.id"` as the body.

Reply:
Given the different nature of the second message (`TokenMessage`) we can't differentiate through a simple flagging byte (or boolean) in the message body without complicating things. Furthermore I can see `TokenMessage` has the metadata field, but `Message` doesn't. I'm not saying it's impossible, just saying that we would introduce unneccessary/ugly logic to achieve something that may not be that useful in the end given the inconsistency above. Just wanted to make you aware of that, let me know what you think

Reply:
Yeah I think that is a fair callout. I could imagine you could discriminate based on  something like message length (given that you don't really need to convey any info), but its debatable whether that's cleaner as well